### PR TITLE
LPS-117760 Improve openToast handling of html content

### PR DESCRIPTION
**Pull request by jbalsas**
This is a revamp of https://github.com/liferay-frontend/liferay-portal/pull/173

It basically does:
- Assume both `title` and `message` passed to `openToast` are HTML
- Remove `titleType` and `messageType` params and usages

Additional boyscouting:
- In `liferay-trash:*` (both `empty` and `undo`). Update markup and remove unnecessary styling. /cc @liferay-echo
- In `message-boards-web`. Properly translates keys. /cc @liferay-lima

**Visuals**

#### Text Title with Text Body
Screen Shot 2020-08-05 at 10 40 16
Screen Shot 2020-08-05 at 10 41 36

#### HTML Title with Text Body
Screen Shot 2020-08-05 at 10 22 39

#### NO Title with HTML Body 🤷 
Screen Shot 2020-08-05 at 10 29 52

#### Text Title with HTML Body
Screen Shot 2020-08-05 at 11 56 52

#### HTML Title with HTML Body
Screen Shot 2020-08-05 at 12 01 15

```javascript
Liferay.Util.openToast({
    title: 'This <em>is</em> html', 
    message: 'Do you believe it? <button class="btn btn-success btn-sm">Shut up and take my money!</button>'
});
```

### CODE REVIEW

**Comment by jbalsas** [comment]
@pat270, @marcoscv-work, @eduardoallegrini, @drakonux... Do those sizes look good to you? I'm not super satisfied with how the alerts break the content to new lines... it feels like we're wasting vertical space in many cases but not sure why...

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js [hunk 1]
```diff
@@ -104,12 +91,10 @@ function openToast({
 	container,
 	containerId,
 	message = '',
-	messageType = TYPES.TEXT,
 	onClick = () => {},
 	onClose = () => {},
 	renderData = DEFAULT_RENDER_DATA,
-	title,
-	titleType = TYPES.TEXT,
+	title = '',
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:97]: Ugh, I added this initially before I checked if it existed... we can get rid of this.

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js [hunk 2]
```diff
@@ -136,15 +121,16 @@ function openToast({
...
-			title={
-				title && (
-					<Text allowHTML={titleType === TYPES.HTML} string={title} />
-				)
-			}
 			variant={variant}
 			{...toastProps}
 		>
-			<Text allowHTML={messageType === TYPES.HTML} string={message} />
+			<div
```
Comment (jbalsas) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:127]: Using a `div` here since we now render the whole content, so it makes no difference what type of element we use. 

Users need to make sure that `message` is an `inline` element. If it isn't, the message will simply jump to the next line. Not the end of the world 🤷

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js [hunk 3]
```diff
@@ -136,15 +121,16 @@ function openToast({
...
+				dangerouslySetInnerHTML={{
+					__html: `${
+						title ? `<strong class="lead">${title}</strong>` : ''
+					}${message}`,
+				}}
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:132]: Code change looks fine to me. One thing though. At the moment I believe most providers of `title` are going to include a trailing whitespace, but that's a bit painful/noisy for them. We could harmlessly include whitespace between `title` and `message` as a convenience here instead. If they already include trailing whitespace in `title`, the browser will collapse it anyway. ie.

```js
title ? `<strong class="lead">${title}</strong> ` : ''
//                                     note:   ^
```

The only thing that I would want to be sure of is this: *are we sure that `message` and `title` are never coming from untrusted user input?*

That's why when [I originally listed](https://github.com/liferay-frontend/liferay-portal/pull/173#issuecomment-668443561) "always assume HTML" as a possibility, I said:

> just assume HTML always (would require an analysis of existing use cases — may be risky)

As long as we're sure that all the messages at all the call sites are coming from us, then 👍

Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:132]: > One thing though. At the moment I believe most providers of title are going to include a trailing whitespace [...]

So, `.lead` already includes `margin-right: 0.3125rem`. I initially had the extra spacey thing and it made the message look off.

> The only thing that I would want to be sure of is this: are we sure that message and title are never coming from untrusted user input?
>
> As long as we're sure that all the messages at all the call sites are coming from us, then 👍

I'm sure they might not, and not sure we can do this at this level. 

I know you usually need to escape the closer to the context that you can, which initially would indeed indicate here which is where we're rendering the html. We don't have the knowledge about whether `message` and `title` are safe though, so I think one idea that was floated around in the past was to mark these as `unsafe` so that consumers of the API would know to escape if needed.

Most messages are simple enough and I don't seem to contain user-generated input on a quick inspection. In general, I see a lot of wrapping, though, so it's hard to track what content is being passed down the chain...

Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:132]: > So, `.lead` already includes `margin-right: 0.3125rem`. I initially had the extra spacey thing and it made the message look off.

So we still have the trailing spaces in the callsites then ([example here](https://github.com/liferay-frontend/liferay-portal/blob/f2a006a711a5ca8e4ee96861f07badc87f6b0a20/portal-web/docroot/html/common/themes/top_js.jspf#L125)), which means we have spacing-in-text _and_ spacing-in-CSS. I think we need the spacing in the text to stop the screenreader from saying something like "titlemessage" when the markup is `<span>title:</span>message`, which is why I suggested having it here once instead of at the callsites. Maybe it only looked weird because you added it here but didn't remove it from the callsites?

> not sure we can do this at this level.

Yeah, I think by definition, when you have an API that takes input and renders HTML, you're already past the point where you can check whether or not you trust the input — it's a string at that point.

In practical terms, I guess we just have to go with your quick inspection, and maybe plaster a warning in the doc comment (that nobody will read) that the message and title are rendered raw and must come from a trusted source. Another option that might help is to rename the parameters to `messageHTML` and `titleHTML`, so at least the caller knows that they are going to be dumped as-is into the DOM; it's not a foolproof safeguard, but it does at least increase the chance that they _might_ think about passing in untrusted input. 🤷

Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:132]: > So we still have the trailing spaces in the callsites then (example here), which means we have spacing-in-text and spacing-in-CSS. I think we need the spacing in the text to stop the screenreader from saying something like "titlemessage" when the markup is <span>title:</span>message, which is why I suggested having it here once instead of at the callsites. Maybe it only looked weird because you added it here but didn't remove it from the callsites?

That's a fair point. I actually remember looking at something similar when we did [Updates clay:alert and clay:stripe tag implementation and usages](https://github.com/brianchandotcom/liferay-portal/pull/90058/files). In that case, was making sure title messages didn't need the `:`.

I checked right now both with `ChromeVox` and `VoiceOver` on mac. 
- They both seem to take the `:` as an almost full stop, so when present, it does read as "title{full_stop}message".
- `VoiceOver` does by default a better job at separating the words. I suspect it might be dependant on their html grouping/containers.
- None seem to react to spaces. Adding any amount of spacing around the colon doesn't seem to change the pause between words (not that I did a proper scientific measurement).

I checked all call sites (kinda) and only a few manually add the `:` and most of them use the usual simple title keys: `succes`, `error`, `warning`... Based on that, and the fact that we already [do this for `clay:alert` and `clay:stripe`](https://github.com/liferay-frontend/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/AlertTag.java#L241), it would seem reasonable to:
- Always add the colon inside `openToast` as we generate the toast content
- Change call sites that are currently passing a colon so they don't pass one (forgivable breaking change, maybe...)
- Add a default `type` to `title` transformation to reduce boilerplate. We [already do this for `clay:alert` and `clay:stripe`](https://github.com/liferay-frontend/liferay-portal/blob/master/modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/AlertTag.java#L270-L281).

Does this seem reasonable to you, @wincent?

> In practical terms, I guess we just have to go with your quick inspection, and maybe plaster a warning in the doc comment (that nobody will read) that the message and title are rendered raw and must come from a trusted source. Another option that might help is to rename the parameters to messageHTML and titleHTML, so at least the caller knows that they are going to be dumped as-is into the DOM; it's not a foolproof safeguard, but it does at least increase the chance that they might think about passing in untrusted input. 🤷

This would be a bigger breaking change though... we could keep both, deprecate the non-suffixed one and replace our usages... not sure how much this is worth, though 🤔

Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:132]: > Does this seem reasonable to you, @wincent?

Yes, although funnily enough, I actually mistyped the example above; instead of:

> saying something like "titlemessage" when the markup is `<span>title:</span>message`

I meant to write:

> saying something like "titlemessage" when the markup is `<span>title</span>message`

The colon snuck in there by mistake. I'm not at all surprised that a screenreader reads "foo:bar" as "foo _pause_ bar".

> not sure how much this is worth, though

I don't either. The probability of this biting us in the form of a nasty security incident in the future is somewhere between 0 and 1, but likely to be closer to the 0 end of the spectrum. And the probability that renaming the param makes the mistake less likely is also somewhere between 0 and 1, but likely closer to the 0 end too... In other words, I think it's ok to leave it as it is, but I at least wanted to float the idea.

Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:132]: Do you think there's some value in adding both `:` and ` ` like `<strong class="lead">${title}:</strong> ${message}`?

Comment (wincent) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:132]: I think if you're going to mess with it (eg. adding colon) you may as well do all the touching at once (ie. adding the space).

**Comment by jbalsas** [comment]
Ok, so in light of @wincent comments, I added:

#### [Generates title by default](https://github.com/liferay-frontend/liferay-portal/pull/208/commits/22da2a5441ab27e5f3e5e0bd08a3291951b12fd5)

This adds:
- Default `title` mapping for the usual `type` values
- Default `:` to the title but no additional space (for now)

```javascript
// success (default) message with "Success:" title
openToast({message: 'foo'});

// danger message with "Error:" title
openToast({message: 'foo', type: 'danger'});

// danger message with no title
openToast({message: 'foo', title: null, type: 'danger'});
openToast({message: 'foo', title: '', type: 'danger'});

// danger message with 'Bar:' title
openToast({message: 'foo', title: 'Bar', type: 'danger'});
```

I tried with and without the additional space and this is how they differ visually:

| space | no space|
| --- | --- |
| Screen Shot 2020-08-05 at 16 56 10 | Screen Shot 2020-08-05 at 16 57 49 |
| Screen Shot 2020-08-05 at 16 56 19 | Screen Shot 2020-08-05 at 16 57 54 |

#### [Removes unnecessary title arguments](https://github.com/liferay-frontend/liferay-portal/pull/208/commits/5e39cfbcc420eac11001d5321a75dc3f491f5d3a)

In here we:
- Remove unnecessary trailing `:`
- Remove redundant `title` values when they match the default mapped values

There's one thing left to decide and it's wether we want to also set `title: ''` to all the occurrences where title wasn't being passed and keep the original (not-sure-how-conscious) behaviour or if we'd rather shake the tree a little to see how many coconuts we get 😂 

@wincent, thoughts?

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by diegonvs** [comment]
I'm fine with all changes but I think we must test for some tricky cases like when showing the dialog which notifies that the session will expire. I remember it does multiple calls of openToast component and I need to test it. I'm building DXP with your fixes.

**Comment by pat270** [comment]
> I'm not super satisfied with how the alerts break the content to new lines...

@jbalsas The lines are breaking prematurely. `.alert-autofit-row { width: auto; }` isn't winning, I need to fix this in Clay CSS.

> I tried with and without the additional space and this is how they differ visually:

The spacing between `.lead` and text is supposed to be 8px from what I remember. The space should be present. I set `.lead { margin-right: 0.3125rem; // 5px }` to account for 3px of white space.

**Comment by jbalsas** [comment]
> The spacing between .lead and text is supposed to be 8px from what I remember. The space should be present. I set .lead { margin-right: 0.3125rem; // 5px } to account for 3px of white space.

That kinda seals it then between me and @wincent 😂 

Isn't it easier to enforce if the whole `8px` are in the margin of the element so we don't rely on a white space that may or may not be added and may or may not be `3px` wide?

**Comment by diegonvs** [comment]
When emulating a session timeout doing: 

```js
AUI().use('liferay-session', () => {
    Liferay.Session = new Liferay.SessionBase(
        {
            autoExtend: false,
            redirectOnExpire: false,
            redirectUrl: 'https://localhost:8080',
            sessionLength: 10,
            warningLength: 10
        }
    );
});
Liferay.Session.plug(Liferay.SessionDisplay);
```

When running the timeout, the counter doesn't update:

Screen Shot 2020-08-05 at 14 41 38

and It suddenly:

Screen Shot 2020-08-05 at 14 41 46

**Comment by pat270** [comment]
@jbalsas It depends on the technology. I think React strips white space where 8px margin makes sense.

If you are using HTML without any post processing you have to intentionally write your markup without white space.
```
<strong class="lead">Primary:</strong><span>The text is here</span>

<strong class="lead">Primary:</strong
><span>The text is here</span>

<strong class="lead">Primary:</strong><!--
--><span>The text is here</span>
```
It can get hard to maintain.

**Comment by diegonvs** [comment]
##  Things to be considered to solve https://github.com/liferay-frontend/liferay-portal/pull/208#issuecomment-669333030:

* Change `session.js` logic to reuse the same openToast container and set the remaining time as the old logic(before `c4233cf0b4615759f5f493c113644d64659603e1`) was doing, querying the dom by className and then changing the node content with the new remainingTime string. 

#### Problem caused

When changing this logic multiple toasts will be rendered(one per second)
containing the updated data.

This is issue was raised because recently we changed the logic to rendering alert to be stacked and in the past when not stacking we had only one toast rendered on the screen.

#### How could we solve it?

##### Option 1:
Changing openToast logic to return a Id, for identify which toast needs to be updated. I’m not sure if we can do an elegant code for it.

##### Option 2:
Add a `stacked` flag to openToast properties so, We can decide if the toast will be stacked or not.

##### Option 3:

Add a new command called `openAlert` for replacing usages that can’t be stacked. AFAIK only this one.

* Rewrite liferay-session for using React since hoisting and controlling a React component

#### Problem caused

Time, deciphering corner cases and test rounds

**Comment by jbalsas** [comment]
Ok... here we go again. From yesterday...
- The `app-builder-web` failure was legit. Pushed small test adjustment to the proper commit (just had to remove the additional `:` in there as well)
- "Fixed" the [session toast thingy](https://github.com/liferay-frontend/liferay-portal/pull/208#issuecomment-669333030)
    - I'm honestly not sure if that ever worked
    - I partially reverted https://github.com/liferay-frontend/liferay-portal/commit/a412961ab982171fc392902a746033c63cf13acd to have a fresh start
    - Provide a kind of [wrapper API](https://github.com/liferay-frontend/liferay-portal/pull/208/commits/d1040d5f8794029548a9993bac62bf6996b8cfc5) to mock the previous one and allow DOM mutation to work the same way. There should be very little changes outside of the `_getBanner` method. 
        - Does it work? Looks like it...
        - Does it work on screen reader? Apparently... (tested with ChromeVox and VoiceOver)
        - Am I proud of this? 🤷 

sel

**Comment by wincent** [comment]
@pat270

> I think React strips white space

It can be hard to remember exactly what it does, but [this page](https://www.javascriptstuff.com/how-whitespace-works-in-jsx/) gives a decent summary; basically:

> on a single line ... HTML and JSX agree on whitespace ... even though JSX renders multiple spaces in your HTML, the browser will still collapse sequences of whitespace. (Assuming that the CSS property `white-space` is left at its default: `normal`).
>
> When we use multiple lines ... the whitespace at the beginning and end of each line is removed, and newlines are removed

@jbalsas 

> There's one thing left to decide and it's wether we want to also set `title: ''` to all the occurrences where title wasn't being passed and keep the original (not-sure-how-conscious) behaviour or if we'd rather shake the tree a little to see how many coconuts we get 😂

By "original (not-sure-how-conscious) behavior", do you mean?

- Show no title at all when none was passed
- Show `<strong className="lead">{title}</strong>` when one passed (via [ClayAlert's title prop](https://github.com/liferay/clay/blob/8521907a9030caf86b06ca579b1eba11333d0e80/packages/clay-alert/src/index.tsx#L163)).
- Which effectively together given a title `'something:'` and message `'yo'` would have rendered `<strong class="lead">something:</span>yo` (note: colon is not injected automatically; it is included by caller).

And new behavior:

- If no title passed, but a default title exists for the toast `type` (which should pretty much always be true, right?), use default.
- If using default or an explicit non-empty title passed, render as `<div><strong class="lead">title:</strong>message</div>` (note: colon is injected automatically).
- If empty string passed as title, don't render title at all, just render `<div>message</div>`.

ie. so you mean that in some places we're going to start showing "Success:" etc based on type where we weren't previously doing it? FWIW, I don't think that matters much: if people object to the change we'll get a bug report and we can fix it easily.

The only thing is that passing `''` to suppress the title is pretty subtle from a type perspective, because it is still a string. I think a better explicit flag for "no value" is `null`.

And the new auto-inject colon behavior is mostly convenient and will produce the desired result, but there is one case I see where we produce a different result (ie. where we had `'<liferay-ui:message key="maintenance-alert" /> <span class="notice-date"><%= FastDateFormatFactoryUtil.getTime(locale).format(Time.getDate(CalendarFactoryUtil.getCalendar(timeZone))) %> <%= timeZone.getDisplayName(false, TimeZone.SHORT, locale) %></span> '` with a space but no colon, and now we'll have a colon. That's probably fine I guess 🤷.

**Comment by wincent** [comment]
☝️ I wrote that without the benefit of seeing your last comment and update @jbalsas, so no idea if it's still relevant or accurate. 😁 

Going to take a look at the code again now (but I've already decided that I like it before looking, 😉 so it's just a formality).

File: modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js [hunk 4]
```diff
@@ -39,7 +39,7 @@ describe('toast', () => {
 			message: Liferay.Language.get(
 				'your-request-completed-successfully'
 			),
-			title: `${Liferay.Language.get('success')}:`,
+			title: `${Liferay.Language.get('success')}`,
```
Comment (wincent) [inline modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js:42]: Unnecessary template literal.

File: modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js [hunk 5]
```diff
@@ -59,7 +59,7 @@ describe('toast', () => {
 
 		expect(openToast).toBeCalledWith({
 			message: Liferay.Language.get('an-unexpected-error-occurred'),
-			title: `${Liferay.Language.get('error')}:`,
+			title: `${Liferay.Language.get('error')}`,
```
Comment (wincent) [inline modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js:62]: Another unnecessary literal.

File: modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/js/utils/toast.es.js [hunk 6]
```diff
@@ -20,7 +20,7 @@ export const errorToast = (
 ) => {
 	openToast({
 		message,
-		title: `${title}:`,
+		title: `${title}`,
```
Comment (wincent) [inline modules/apps/data-engine/data-engine-taglib/src/main/resources/META-INF/resources/data_layout_builder/js/utils/toast.es.js:23]: Two more in this file that could use shorthand (also line 34):

```js
title,
```

File: modules/apps/dynamic-data-lists/dynamic-data-lists-web/src/main/resources/META-INF/resources/view_records.jsp [hunk 7]
```diff
@@ -201,9 +201,9 @@ if (!ddlDisplayContext.isAdminPortlet()) {
 		deleteRecords: deleteRecords,
 	};
 
-	Liferay.componentReady('<%= randomNamespace + "ddlViewRecordsManagementToolbar" %>').then(function (
-		managementToolbar
-	) {
+	Liferay.componentReady(
+		'<%= randomNamespace + "ddlViewRecordsManagementToolbar" %>'
+	).then(function (managementToolbar) {
```
Comment (wincent) [inline modules/apps/dynamic-data-lists/dynamic-data-lists-web/src/main/resources/META-INF/resources/view_records.jsp:206]: Sneaking in some formatting fixes eh?

File: modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js [hunk 8]
```diff
@@ -94,7 +94,7 @@ AUI.add(
 
 					instance.set('timestamp');
 
-					if (event.src === SRC) {
+					if (event.src == SRC) {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js:97]: Intended? We [prefer triple equals anyway](https://github.com/liferay/eslint-config-liferay/issues/158).

Same further down in this file in various places.

File: modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js (hunk 4 above)
Comment (jbalsas) [reply modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js:42]: I love when you get picky... I usually means we're getting closer 😂

looks_wrong

File: modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js (hunk 8 above)
Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js:97]: This was an automated revert and SF did not attempt a fix here... don't mind changing it... 🤷

File: modules/apps/dynamic-data-lists/dynamic-data-lists-web/src/main/resources/META-INF/resources/view_records.jsp (hunk 7 above)
Comment (jbalsas) [reply modules/apps/dynamic-data-lists/dynamic-data-lists-web/src/main/resources/META-INF/resources/view_records.jsp:206]: Not sure :D

File: modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js [hunk 9]
```diff
@@ -538,58 +559,76 @@ AUI.add(
...
-							messageType: 'html',
-							onClick(eventPayload) {
-								var event = eventPayload.event;
-								var onClose = eventPayload.onClose;
+					if (!banner) {
+						var openToast = instance.get('openToast');
+						var toastId = 'sessionToast';
 
+						var toastDefaultConfig = {
+							onClick: ({event}) => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js:572]: Pro-Tip™: totally optional, but you could use object concide methods here (value of `this` not needed, so no need for lexical scoping of `this` provided by the arrow). ie.

```js
var toastDefaultConfig = {
    onClick({event}) {
        // stuff...
    },    
    onClose({event}) {
        // blah...
    },
    // etc
};
```

Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js:572]: ugh, yeah, you're right

File: modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js [hunk 10]
```diff
@@ -538,58 +559,76 @@ AUI.add(
...
+							one: A.one(`#${toastId}`).one,
+							setAttrs: (attrs) => {
+								toastComponent.destroy();
+
+								openToast({
+									...attrs,
+									...toastDefaultConfig,
+								});
+							},
+							show: () => {
```
Comment (wincent) [inline modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js:614]: Two more places that would look slightly neater/briefer with object concise methods.

File: modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js (hunk 4 above)
Comment (wincent) [reply modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js:42]: Next step is:

> Can you break this into two parts and resend?

File: modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js (hunk 8 above)
Comment (wincent) [reply modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js:97]: Glad to hear that SF didn't break it (we still need to turn on the lint, but every `==` we add to the codebase makes it harder to turn on the lint, which is why I always comment on them when I see them.)

Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js:97]: Yeah, I agree. This seems brittle enough that I don't want to touch anything else than necessary. I think we can revamp the whole thing after `7.3`. This is a huge source of issues and it's still AUI, so...

just-sayin

File: modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js (hunk 4 above)
Comment (jbalsas) [reply modules/apps/app-builder/app-builder-web/test/js/utils/toast.es.js:42]: Don't see why not! :)

File: modules/apps/dynamic-data-lists/dynamic-data-lists-web/src/main/resources/META-INF/resources/view_records.jsp (hunk 7 above)
Comment (jbalsas) [reply modules/apps/dynamic-data-lists/dynamic-data-lists-web/src/main/resources/META-INF/resources/view_records.jsp:206]: I'm going to look the other way with this one... let it sneak in... like a ninja...

File: modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js (hunk 10 above)
Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-aui-web/src/main/resources/META-INF/resources/liferay/session.js:614]: It's going to look beautiful!

**Comment by jbalsas** [comment]
> perfect

Based on this, I basically just changed what I felt changing and left unchanged what I didn't feel like changing. Hope that's a good summary of what I did.

> The only thing is that passing '' to suppress the title is pretty subtle from a type perspective, because it is still a string. I think a better explicit flag for "no value" is null.

Yup, `null` also works to negate the existence of `title` if we have to go for it.

> ie. so you mean that in some places we're going to start showing "Success:" etc based on type where we weren't previously doing it? FWIW, I don't think that matters much: if people object to the change we'll get a bug report and we can fix it easily.

Exactly, let's wing it!

> And the new auto-inject colon behavior is mostly convenient and will produce the desired result, but there is one case I see where we produce a different result (ie. where we had '<liferay-ui:message key="maintenance-alert" /> <span class="notice-date"><%= FastDateFormatFactoryUtil.getTime(locale).format(Time.getDate(CalendarFactoryUtil.getCalendar(timeZone))) %> <%= timeZone.getDisplayName(false, TimeZone.SHORT, locale) %></span> ' with a space but no colon, and now we'll have a colon. That's probably fine I guess 🤷.

Yeah, that's what I thought too... that's why I marked it as a `forgivable breaking change` in some comment above...

**Comment by jbalsas** [comment]
ci:test:sf

**Comment by jbalsas** [comment]
ci:test:relevant

**Comment by wincent** [comment]
> I basically just changed what I felt changing and left unchanged what I didn't feel like changing.

deal-with-it

**Comment by jbalsas** [comment]
❌ ci:test:stable - 20 out of 21 jobs passed
❌ ci:test:relevant - 185 out of 192 jobs passed in 3 hours 3 minutes

Sounds reasonable... of the failures:
- Semver (unrelated)
- Other stuff (they all sport toasts in the screen but I don't think any is asserting anything about them)

@john-co, mind confirming quickly before I manually forward this?

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js [hunk 11]
```diff
@@ -123,11 +115,13 @@ function openToast({
 			onClose({event});
 		}
 
-		if (!container || !containerId) {
-			rootElement.parentNode.removeChild(rootElement);
-		}
+		if (!event.defaultPrevented) {
```
Comment (diegonvs) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:118]: Hey @jbalsas, when opening the session toast and triggering another toast by typing on console `Liferay.Util.openToast({message: 'test'})`, an error will be thrown: 

Screen Shot 2020-08-06 at 14 07 47

I think we must just make sure We have `event` defined here

Comment (jbalsas) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/toast/commands/OpenToast.es.js:118]: Hey @diegonvs, that's a nice catch! I'm curious... did you actually see an error occur?

The only way I see this can happen is if you actually created an `onClick` handler that called the `onClose` method without passing an `event` (which is not the snippet you pasted, thus the curiosity)

```javascript
Liferay.Util.openToast({
    message: ' message hey you', 
    onClick({event, onClose}) {onClose();} 
});
```

I'll if the thing off and manually push to Brian, see if we make it to the release window.

**Comment by jbalsas** [comment]
As promised: https://github.com/brianchandotcom/liferay-portal/pull/92355

**Comment by john-co** [comment]
ci:test:bundle

**Comment by john-co** [comment]
> Sounds reasonable... of the failures:
> 
> Semver (unrelated)
> Other stuff (they all sport toasts in the screen but I don't think any is asserting anything about them)

This sounds about right. Reviewed. Test failures are unrelated. :heavy_check_mark:

**Comment by john-co** [comment]
Hi @diegonvs , looks like this caused https://issues.liferay.com/browse/LPS-118940
